### PR TITLE
Support segment upload with Uri

### PR DESCRIPTION
**Pull request by fx19880617**
Adding a default maxRetries for HttpSegmentFetcher to 3.
Adding enum type FileUploadType to support Tar and Uri.
Tar is the default one which is currently using.
For Uri mode, instead of curling the entire segment in a POST request, we just put segment download uri in the request header and controller will fetch segment to local based on the given uri.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -26,6 +26,8 @@ target/
 .settings/
 out/
 .DS_Store
+bin/
```
Comment (jfim) [inline .gitignore]: Can you make this a different commit?

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java [hunk 2]
```diff
@@ -26,15 +26,35 @@
...
-    LOGGER.info("Downloaded file from {} to {}; Length of httpGetResponseContent: {}; Length of downloaded file: {}", uri, tempFile,
-        httpGetResponseContentLength, tempFile.length());
+    for (int retry = 1; retry <= _maxRetries; ++retry) {
+      try {
+        final long httpGetResponseContentLength = FileUploadUtils.getFile(uri, tempFile);
+        LOGGER.info("Downloaded file from {} to {}; Length of httpGetResponseContent: {}; Length of downloaded file: {}", uri, tempFile,
+            httpGetResponseContentLength, tempFile.length());
+        return;
+      } catch (Exception e) {
+        LOGGER.error(String.format("Failed to download file from %s, retry: %d", uri, retry), e);
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: Use `LOGGER.error("Failed to download file from {}, attempt {}, e, uri, retry)`

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java [hunk 3]
```diff
@@ -107,6 +109,32 @@ public static int sendSegmentFile(final String host, final String port, final St
...
+
+  public static int sendSegmentUri(final String host, final String port, final String uri) {
+    HttpClient client = new HttpClient();
+    SendFileMethod httpMethod = SendFileMethod.POST;
+    try {
+      client.getParams().setParameter("http.protocol.version", HttpVersion.HTTP_1_1);
+      client.getParams().setSoTimeout(3600 * 1000); // One hour
+      EntityEnclosingMethod method = httpMethod.forUri("http://" + host + ":" + port + "/" + SEGMENTS_PATH);
+      method.setRequestHeader(UPLOAD_TYPE, FileUploadType.Uri.toString());
+      method.setRequestHeader(DOWNLOAD_URI, uri);
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: It would be a lot better if we uploaded a JSON object with instructions as opposed to doing an empty post with a HTTP header.

File: .gitignore (hunk 1 above)
Comment (fx19880617) [inline .gitignore]: will do

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java (hunk 2 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: will do

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java (hunk 3 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: if we post a json file, how do u know it's a json not the entire segment tar?

File: .gitignore (hunk 1 above)
Comment (fx19880617) [inline .gitignore]: https://github.com/linkedin/pinot/pull/130

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/broker/BrokerServerBuilder.java [hunk 4]
```diff
@@ -244,8 +244,12 @@ public void start() throws Exception {
 
   public void stop() throws Exception {
     LOGGER.info("Shutting down Network !!");
-
     _state.set(State.SHUTTING_DOWN);
+    try {
+      Thread.sleep(10000);
```
Comment (raviarin) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/broker/BrokerServerBuilder.java]: What is the need for the sleep if all we do is close connection pool and shutdown?

Comment (fx19880617) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/broker/BrokerServerBuilder.java]: Mostly for both health check and externalView to be in offline status.
I've observed many cases when I'm bouncing broker, the client side with see a short period that query response is empty.
I tried to setup both haproxy pinging health check  and also register an externalView listener. Since we are stopping too fast, either of this approach will treat this broker is still alive, but actually getting empty results.

The clean way here is to set idealstates of this broker to offline meanwhile still serving queries. Then  wait for externalView's reflection to offline then shutdown broker.

Also a configurable fixed timeout is necessary if I'm using haproxy to do health check.

Comment (antumbde) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/broker/BrokerServerBuilder.java]: Broker will continue to server requests for 10seconds in this case ? Will this addressing the issue in that case ?

It also seems unrelated to the main PR. Move to a separate PR

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java [hunk 5]
```diff
@@ -26,15 +26,35 @@
 public class HttpSegmentFetcher implements SegmentFetcher {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(HttpSegmentFetcher.class);
+  private static final String MAX_RETRIES = "maxRetries";
+  private static final int DEFAULT_MAX_RETRIES = 3;
+  private int _maxRetries = DEFAULT_MAX_RETRIES;
```
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: don't use '_' prefixed variables. We are following google coding convention

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java [hunk 6]
```diff
@@ -26,15 +26,35 @@
...
-    LOGGER.info("Downloaded file from {} to {}; Length of httpGetResponseContent: {}; Length of downloaded file: {}", uri, tempFile,
-        httpGetResponseContentLength, tempFile.length());
+    for (int retry = 1; retry <= _maxRetries; ++retry) {
+      try {
+        final long httpGetResponseContentLength = FileUploadUtils.getFile(uri, tempFile);
+        LOGGER.info("Downloaded file from {} to {}; Length of httpGetResponseContent: {}; Length of downloaded file: {}", uri, tempFile,
+            httpGetResponseContentLength, tempFile.length());
+        return;
+      } catch (Exception e) {
+        LOGGER.error("Failed to download file from {}, retry: {}", e, uri, retry);
```
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: exception should be last param

Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: better to backoff instead of retrying in a tight loop

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java [hunk 7]
```diff
@@ -107,6 +113,60 @@ public static int sendSegmentFile(final String host, final String port, final St
     return sendFile(host, port, SEGMENTS_PATH, fileName, inputStream, lengthInBytes, SendFileMethod.POST);
   }
 
+
+  public static int sendSegmentUri(final String host, final String port, final String uri) {
+    HttpClient client = new HttpClient();
+    SendFileMethod httpMethod = SendFileMethod.POST;
+    try {
+      client.getParams().setParameter("http.protocol.version", HttpVersion.HTTP_1_1);
+      client.getParams().setSoTimeout(3600 * 1000); // One hour
```
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: make this and one above static constant

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java [hunk 8]
```diff
@@ -136,4 +196,26 @@ public static long getFile(String url, File file) throws Exception {
       throw ex;
     }
   }
+
+  public enum FileUploadType {
+    Uri,
```
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: all caps

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java (hunk 3 above)
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: +1 to jfim's comment. 
Send a json object that allows switching between json and URI versions.
Sending data as headers is very odd

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java [hunk 9]
```diff
@@ -107,6 +113,60 @@ public static int sendSegmentFile(final String host, final String port, final St
...
+    SendFileMethod httpMethod = SendFileMethod.POST;
+    try {
+      client.getParams().setParameter("http.protocol.version", HttpVersion.HTTP_1_1);
+      client.getParams().setSoTimeout(3600 * 1000); // One hour
+      EntityEnclosingMethod method = httpMethod.forUri("http://" + host + ":" + port + "/" + SEGMENTS_PATH);
+      method.setRequestHeader(UPLOAD_TYPE, FileUploadType.Uri.toString());
+      method.setRequestHeader(DOWNLOAD_URI, uri);
+      client.executeMethod(method);
+      if (method.getStatusCode() >= 400) {
+        String errorString = "POST Status Code: " + method.getStatusCode() + "\n";
```
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: handle 3xx here and below

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java [hunk 10]
```diff
@@ -107,6 +113,60 @@ public static int sendSegmentFile(final String host, final String port, final St
...
+        if (method.getResponseHeader("Error") != null) {
+          errorString += "ServletException: " + method.getResponseHeader("Error").getValue();
+        }
+        throw new HttpException(errorString);
+      }
+      return method.getStatusCode();
+    } catch (Exception e) {
+      LOGGER.error("Caught exception while sending file", e);
```
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: log filename

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java [hunk 11]
```diff
@@ -107,6 +113,60 @@ public static int sendSegmentFile(final String host, final String port, final St
...
+      if (statusCode >= 400) {
+        String errorString = "POST Status Code: " + statusCode + "\n";
+        if (postMethod.getResponseHeader("Error") != null) {
+          errorString += "ServletException: " + postMethod.getResponseHeader("Error").getValue();
+        }
+        throw new HttpException(errorString);
+      }
+      return statusCode;
+    } catch (Exception e) {
+      LOGGER.error("Caught exception while sending file", e);
```
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: log filename that had error

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java [hunk 12]
```diff
@@ -215,30 +224,81 @@ public Representation post(Representation entity) {
...
+        uploadType = (uploadTypeStr == null) ? FileUploadType.getDefaultUploadType() : FileUploadType.valueOf(uploadTypeStr);
+      } catch (Exception e) {
+        uploadType = FileUploadType.getDefaultUploadType();
+      }
+      String downloadURI = null;
       boolean found = false;
-      for (final Iterator<FileItem> it = items.iterator(); it.hasNext() && !found;) {
-        final FileItem fi = it.next();
-        if (fi.getFieldName() != null) {
+      switch (uploadType) {
```
Comment (antumbde) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java]: fix indentation

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java [hunk 13]
```diff
@@ -215,30 +224,81 @@ public Representation post(Representation entity) {
...
+      case Uri:
+      case Json:
+        // Download segment from the given Uri
+        try {
+          downloadURI = getDownloadUri(uploadType, headers, entity);
+        } catch (Exception e) {
+          String errorMsg = String.format("Failed to get download Uri for upload file type: %s, with error %s", uploadType, e.getMessage());
+          LOGGER.warn(errorMsg);
+          ControllerRestApplication.getControllerMetrics().addMeteredGlobalValue(ControllerMeter.CONTROLLER_SEGMENT_UPLOAD_ERROR, 1L);
+          setStatus(Status.SERVER_ERROR_INTERNAL);
```
Comment (antumbde) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java]: send json error message
makes it easier to add structure to error body

Comment (antumbde) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java]: this is bad request and not server error

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java [hunk 14]
```diff
@@ -292,14 +352,30 @@ public Representation post(Representation entity) {
...
+  }
+
   @HttpVerb("post")
   @Summary("Uploads a segment")
   @Tags({"segment"})
   @Paths({
       "/segments",
       "/segments/"
   })
-  private Representation uploadSegment(File indexDir, File dataFile)
```
Comment (antumbde) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java]: Document API and add UTs

**Comment by antumbde** [comment]
This needs API documentation and unit tests

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/broker/BrokerServerBuilder.java (hunk 4 above)
Comment (fx19880617) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/broker/BrokerServerBuilder.java]: + 10 secs will solve the issue from my observation.
Will make a different pr for this.

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java (hunk 7 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: done

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java (hunk 5 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: done

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java (hunk 6 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: done

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java (hunk 8 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: done

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java (hunk 10 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: done

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java (hunk 11 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: done

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java (hunk 13 above)
Comment (fx19880617) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java]: done for bad request.

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java (hunk 3 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: implemented upload types for uri/json/tar.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java (hunk 12 above)
Comment (fx19880617) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java]: From https://github.com/linkedin/pinot/blob/master/pinot-style.xml, there is not indentation for switch/case.

Comment (fx19880617) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java]: Done.
I think this part is out-dated, we may need to update pinot-style.xml.

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java [hunk 15]
```diff
@@ -43,6 +47,14 @@
 
   private static final Logger LOGGER = LoggerFactory.getLogger(FileUploadUtils.class);
   private static final String SEGMENTS_PATH = "segments";
+  public static final String UPLOAD_TYPE = "UPLOAD_TYPE";
+  public static final String DOWNLOAD_URI = "DOWNLOAD_URI";
+  private static final HttpClient FILE_UPLOAD_HTTP_CLIENT = new HttpClient();
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: If the same instance of HttpClient is reused, then a multithread-aware connection manager should be used. See http://hc.apache.org/httpclient-3.x/threading.html

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java [hunk 16]
```diff
@@ -26,15 +26,38 @@
 public class HttpSegmentFetcher implements SegmentFetcher {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(HttpSegmentFetcher.class);
+  private static final String MAX_RETRIES = "maxRetries";
+  private static final int DEFAULT_MAX_RETRIES = 3;
+  private int maxRetries = DEFAULT_MAX_RETRIES;
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: `maxRetryCount`

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java [hunk 17]
```diff
@@ -26,15 +26,38 @@
...
+    for (int retry = 1; retry <= maxRetries; ++retry) {
+      try {
+        final long httpGetResponseContentLength = FileUploadUtils.getFile(uri, tempFile);
+        LOGGER.info("Downloaded file from {} to {}; Length of httpGetResponseContent: {}; Length of downloaded file: {}", uri, tempFile,
+            httpGetResponseContentLength, tempFile.length());
+        return;
+      } catch (Exception e) {
+        LOGGER.error("Failed to download file from {}, retry: {}", uri, retry, e);
+        if (retry == maxRetries) {
+          throw e;
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: Add logging statement "Exceeded maximum retry count while fetching {}, aborting."

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java [hunk 18]
```diff
@@ -26,15 +26,38 @@
...
+      }
+    }
   }
 
   @Override
   public void fetchSegmentToLocal(String uri, File tempFile) throws Exception {
-    final long httpGetResponseContentLength = FileUploadUtils.getFile(uri, tempFile);
-    LOGGER.info("Downloaded file from {} to {}; Length of httpGetResponseContent: {}; Length of downloaded file: {}", uri, tempFile,
-        httpGetResponseContentLength, tempFile.length());
+    for (int retry = 1; retry <= maxRetries; ++retry) {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: When we get a segment refresh message from the controller, we do not want to retry multiple times. That will hold up other segment refresh messages.

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java (hunk 17 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: Updated.

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java (hunk 16 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: updated

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java (hunk 15 above)
Comment (fx19880617) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: updated

**Comment by jfim** [comment]
lgtm, I'll merge it once travis finishes

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java (hunk 18 above)
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: The segment refresh message IS asynchronous. If we handle multiple refresh messages at the same time, we can run out of memory on the servers, since a refreshed segment is not unloaded until all the queries are completed for that segment.

Further, the entire helix pipeline is not blocked on a refresh. Only other segment refreshes are blocked. The rest of the helix transition, etc. will happen as usual.

Ideally, we need to differentiate between a download failure vs load failure (exception on load). The former is intermittent, and the latter is probably permanent, and there is no reason to retry.

In that sense, adding a retry for HttpFetcher (for the server part) seems fine. We need to change the logic that calls this class and handle  things appropriately.
